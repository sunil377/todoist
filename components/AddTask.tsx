import { db } from '@/config/firebase'
import { FirebaseError } from 'firebase/app'
import { addDoc, collection, serverTimestamp } from 'firebase/firestore'
import { useFormik } from 'formik'
import { parseZodError } from 'helpers/util'
import isEmpty from 'lodash/isEmpty'
import InboxIcon from 'public/assets/inbox.svg'
import TodayIcon from 'public/assets/today.svg'
import { useMutation } from 'react-query'
import { z } from 'zod'

const INITIAL_VALUES = {
    title: '',
    description: '',
}

function useTask() {
    const formik = useFormik({
        initialValues: INITIAL_VALUES,
        validateOnMount: true,
        validate: validate,
        onSubmit: () => {
            mutation.mutate()
        },
    })

    const mutation = useMutation(
        () =>
            addDoc(collection(db, 'tasks'), {
                ...formik.values,
                createdAt: serverTimestamp(),
            }),
        {
            onError: (err) => {
                if (err instanceof FirebaseError) {
                    formik.setFieldError('err', err.code)
                    return
                }

                console.error('Error happen during adding Task', err)
            },

            onSuccess: (data) => {
                console.log('return taskid after adding task', data.id)
                formik.resetForm()
            },

            onSettled: () => {
                formik.setSubmitting(false)
            },
        },
    )

    return formik
}

function AddTask({ onClose }: { onClose: () => void }) {
    const {
        handleSubmit,
        errors,
        isSubmitting,
        isValid,
        touched,
        getFieldProps,
    } = useTask()

    return (
        <form onSubmit={handleSubmit} noValidate>
            <div className="flex flex-col gap-y-2 rounded-lg border border-gray-300 px-4 py-3 text-sm focus-within:border-gray-400">
                <input
                    type="text"
                    placeholder="Task name"
                    className="outline-none"
                    autoComplete="off"
                    autoCapitalize="none"
                    aria-invalid={!!errors.title && touched.title}
                    aria-labelledby={
                        !!errors.title && touched.title
                            ? 'task-alert'
                            : undefined
                    }
                    required
                    {...getFieldProps('title')}
                />

                <input
                    type="text"
                    placeholder="Description"
                    className="outline-none"
                    autoComplete="off"
                    autoCapitalize="none"
                    aria-invalid={!!errors.description && touched.description}
                    aria-labelledby={
                        !!errors.description && touched.description
                            ? 'task-alert'
                            : undefined
                    }
                    required
                    {...getFieldProps('description')}
                />

                <section className="mt-2 flex gap-x-2" role="group">
                    <button
                        type="button"
                        className="gap-x-1 border-green-300 px-1.5 py-0.5 text-xs font-normal text-green-700"
                    >
                        <TodayIcon aria-hidden />
                        Today
                    </button>

                    <button
                        type="button"
                        className="gap-x-1 border-blue-300 px-1.5 py-0.5 text-xsm font-normal text-blue-700"
                    >
                        <InboxIcon aria-hidden className="text-blue-500" />
                        Inbox
                    </button>
                </section>
            </div>

            <div className="mt-2 flex justify-end gap-x-3 text-xsm">
                <button
                    type="button"
                    className="bg-gray-200 px-2 py-1.5 text-gray-800"
                    onClick={onClose}
                >
                    Cancel
                </button>

                <button
                    type="submit"
                    className="bg-skin-dark px-2 py-1.5 text-white"
                    disabled={isSubmitting || !isValid}
                >
                    Add task
                </button>
            </div>

            {!isValid && !isEmpty(touched) && (
                <p
                    role="alert"
                    className="text-sm text-red-500"
                    id="task-alert"
                >
                    &#42;{Object.values(errors)[0]}
                </p>
            )}
        </form>
    )
}

export default AddTask

function validate<T>(values: T) {
    const response = z
        .object({
            title: z.string().min(3, 'Title should have minimun 3 characters'),
            description: z
                .string()
                .min(3, 'Description should have minimun 3 characters'),
        })
        .safeParse(values)

    return parseZodError(response)
}
