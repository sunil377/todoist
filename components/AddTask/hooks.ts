import { useAuth } from 'context/AuthContext'
import { startOfDay } from 'date-fns'
import { FirebaseError } from 'firebase/app'
import {
    addDoc,
    onSnapshot,
    serverTimestamp,
    updateDoc,
} from 'firebase/firestore'
import { FormikHelpers } from 'formik'
import {
    getProjectCollectionRef,
    getTaskCollectionRef,
    getTaskRef,
} from 'hooks/services'
import { IProject, ITask } from 'index'
import { useEffect, useState } from 'react'
import { useMutation } from 'react-query'

type IState = Pick<ITask, 'description' | 'project' | 'title'> & {
    dueDate: string
}

type mutationType = {
    values: IState
    helpers: FormikHelpers<IState>
}

function useToCreateTask() {
    const currentUser = useAuth()

    return useMutation(
        ({ values: { dueDate, ...rest } }: mutationType) => {
            if (!currentUser) {
                return Promise.reject(new Error('Not Allowed'))
            }

            const d = startOfDay(Date.parse(dueDate)).getTime()

            const data: Omit<ITask, 'updatedAt' | 'id'> = {
                description: rest.description,
                title: rest.title,
                dueDate: d,
                createdAt: serverTimestamp(),
                project: rest.project,
                completed: false,
            }

            return addDoc(getTaskCollectionRef(currentUser.uid), data)
        },
        {
            onSuccess: (_data, { helpers: { resetForm } }) => {
                resetForm()
            },
            onError: (err, { helpers: { setFieldError } }) => {
                setFieldError('other', (err as FirebaseError | Error).message)
            },
            onSettled: (_data, _error, { helpers: { setSubmitting } }) => {
                setSubmitting(false)
            },
        },
    )
}

type updateinitialValues = Pick<ITask, 'description' | 'project' | 'title'> & {
    dueDate: string
}

type updatemutationType = {
    values: updateinitialValues
    helpers: FormikHelpers<updateinitialValues>
}

function useToUpdateTask(docId: string, onSuccess: () => void) {
    const currentUser = useAuth()

    return useMutation(
        ({ values: { dueDate, ...rest } }: updatemutationType) => {
            if (!currentUser) {
                return Promise.reject(new Error('not allowed'))
            }
            const data = {
                ...rest,
                ...(dueDate
                    ? { dueDate: startOfDay(Date.parse(dueDate)).getTime() }
                    : {}),

                updatedAt: serverTimestamp(),
            }
            return updateDoc(getTaskRef(currentUser.uid, docId), data)
        },
        {
            onSuccess: () => {
                onSuccess()
            },

            onSettled: (_data, _error, { helpers: { setSubmitting } }) => {
                setSubmitting(false)
            },
        },
    )
}

function usePicker() {
    const [query, setQuery] = useState('')
    const [tags, setTags] = useState<Array<IProject>>([])
    const currentUser = useAuth()

    const filterTags = tags.filter(({ title }) =>
        title.toLowerCase().includes(query.toLowerCase()),
    )

    useEffect(() => {
        if (!currentUser) {
            return
        }
        return onSnapshot(getProjectCollectionRef(currentUser.uid), (res) => {
            let p: Array<IProject> = []
            res.forEach((t) => {
                if (t.exists()) {
                    const result = { id: t.id, ...t.data() } as IProject
                    p = [result, ...p]
                }
            })
            setTags(p)
        })
    }, [currentUser])

    return {
        query,
        setQuery,
        filterTags,
    }
}

export { useToCreateTask, usePicker, useToUpdateTask }
