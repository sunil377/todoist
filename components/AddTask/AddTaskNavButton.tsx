import { Dialog } from '@headlessui/react'
import { useKeypadListener } from 'hooks/useKeypadListener'
import { Fragment, useCallback, useState } from 'react'
import { MdAdd } from 'react-icons/md'
import ToolTip from '../Tooltip'

import { AddTaskComponent } from './AddTask'

function AddTaskNavButton() {
    const [isOpen, setOpen] = useState(false)

    const onkeydown = useCallback((e: KeyboardEvent) => {
        if (e.key === 'q') {
            e.preventDefault()
            setOpen((prev) => !prev)
        }
    }, [])

    useKeypadListener(onkeydown)

    return (
        <Fragment>
            <button
                className="tooltip relative rounded-full p-2 hover:bg-white/10 focus:outline-none focus-visible:bg-white/20"
                onClick={() => setOpen(true)}
            >
                <ToolTip className="top-full left-1/2 -translate-x-1/2 translate-y-2 after:-top-1 after:left-1/2 after:-translate-x-1/2">
                    Add Task <kbd>Q</kbd>
                </ToolTip>
                <MdAdd className="text-xl md:text-2xl" />
            </button>

            <Dialog
                onClose={setOpen}
                open={isOpen}
                className="fixed inset-x-0 top-1/4 isolate z-50 mx-auto w-10/12 max-w-md"
            >
                <div aria-hidden className="fixed inset-0 z-10 bg-black/20" />
                <Dialog.Panel className="relative z-20 rounded border bg-white p-5">
                    <Dialog.Title className="sr-only">Add task</Dialog.Title>
                    <AddTaskComponent
                        close={
                            <button
                                onClick={() => setOpen(false)}
                                type="button"
                                className="flex-shrink-0 rounded border-4 border-transparent py-1 px-2 text-sm text-gray-500 hover:text-gray-800"
                            >
                                Cancel
                            </button>
                        }
                    />
                </Dialog.Panel>
            </Dialog>
        </Fragment>
    )
}
export default AddTaskNavButton
