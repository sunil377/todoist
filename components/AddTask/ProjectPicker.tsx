import { Combobox, Popover } from '@headlessui/react'
import clsx from 'clsx'
import { useField, useFormikContext } from 'formik'
import { findColor } from 'layout/SidePanel/FooterPanel'
import { Fragment } from 'react'
import { FaInbox } from 'react-icons/fa'
import { HiChevronDown } from 'react-icons/hi'
import { MdCheck } from 'react-icons/md'
import { usePicker } from './hooks'

function ProjectPicker({ isDialog = false }) {
    const { query, setQuery, filterTags } = usePicker()

    const [field] = useField('project')
    const { setFieldValue } = useFormikContext()
    console.log(field)

    return (
        <Popover as="div" className="relative">
            <Popover.Button
                type="button"
                className={clsx(
                    'inline-flex items-center gap-x-1 rounded border border-blue-300 bg-white px-2.5 py-1.5 capitalize focus:outline-none focus-visible:border-blue-300 focus-visible:ring focus-visible:ring-blue-300 focus-visible:ring-offset-2',
                    {
                        'ring-offset-gray-800': isDialog,
                    },
                )}
            >
                {field.value.toLowerCase() === 'inbox' && (
                    <FaInbox aria-hidden className="text-base text-blue-500" />
                )}
                {field.value}
                <HiChevronDown aria-label="down arrow" className="text-base" />
            </Popover.Button>

            <Popover.Panel className="absolute right-0 top-[calc(100%+0.5rem)] w-[240px] overflow-y-auto rounded-md border bg-white text-xsm shadow-md">
                {({ close }) => (
                    <Combobox
                        value={field.value}
                        onChange={(e) => {
                            setFieldValue('project', e)
                            setQuery('')
                            close()
                        }}
                        name={field.name}
                    >
                        <Combobox.Input
                            onChange={(e) => setQuery(e.target.value)}
                            value={query}
                            className="w-full py-2 pl-3 pr-2 focus:outline-none"
                            autoFocus
                            placeholder="Type a project"
                        />

                        <Combobox.Options static className="divide-y border-t">
                            {filterTags.map((tag) => (
                                <Combobox.Option
                                    key={tag.id}
                                    value={tag.title}
                                    as={Fragment}
                                >
                                    {({ active, selected }) => (
                                        <li
                                            className={clsx(
                                                'flex items-center gap-x-2 py-2 px-3 capitalize',
                                                { 'bg-gray-100': active },
                                            )}
                                        >
                                            {tag.title === 'inbox' ? (
                                                <FaInbox
                                                    aria-hidden
                                                    className="text-base text-blue-500"
                                                />
                                            ) : (
                                                <span
                                                    className="inline-block h-3 w-3 rounded-full"
                                                    style={{
                                                        background: findColor(
                                                            tag.color,
                                                        ),
                                                    }}
                                                ></span>
                                            )}
                                            {tag.title}
                                            {selected && (
                                                <MdCheck
                                                    aria-label="selected"
                                                    className="ml-auto text-blue-500"
                                                />
                                            )}
                                        </li>
                                    )}
                                </Combobox.Option>
                            ))}
                        </Combobox.Options>
                    </Combobox>
                )}
            </Popover.Panel>
        </Popover>
    )
}
export default ProjectPicker
