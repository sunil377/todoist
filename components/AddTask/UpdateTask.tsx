import { useAuth } from 'context/AuthContext'
import { format, startOfDay } from 'date-fns'
import { serverTimestamp, updateDoc } from 'firebase/firestore'
import {
    ErrorMessage,
    Field,
    FieldProps,
    Form,
    Formik,
    FormikHelpers,
} from 'formik'
import { parseZodError } from 'helpers/util'
import { getTaskRef } from 'hooks/services'
import { ITask } from 'index'
import { useRef } from 'react'
import { FaInbox } from 'react-icons/fa'
import { MdCalendarToday } from 'react-icons/md'
import { useMutation } from 'react-query'
import { z } from 'zod'

type initialValues = Omit<
    ITask,
    keyof Pick<ITask, 'id' | 'createdAt' | 'dueDate' | 'project'>
> & {
    dueDate: string
}

type mutationType = {
    values: initialValues
    helpers: FormikHelpers<initialValues>
    currentUser: string
}

function UpdateTask(
    props: { handleClose: () => void } & Omit<
        ITask,
        keyof Pick<ITask, 'createdAt' | 'project'>
    >,
) {
    const { handleClose, id, ...initialValues } = props
    const initialState = useRef({
        ...initialValues,
        dueDate: format(new Date(initialValues.dueDate), 'yyyy-MM-dd'),
    })

    const currentUser = useAuth()

    const mutation = useMutation(
        ({ values: { dueDate, ...rest }, currentUser }: mutationType) => {
            const d = startOfDay(Date.parse(dueDate)).getTime()

            const data = {
                ...rest,
                dueDate: d,
                updatedAt: serverTimestamp(),
            }
            return updateDoc(getTaskRef(currentUser, id), data)
        },
        {
            onSuccess: () => {
                handleClose()
            },

            onSettled: (_data, _error, { helpers: { setSubmitting } }) => {
                setSubmitting(false)
            },
        },
    )

    return (
        <>
            <Formik
                initialValues={initialState.current}
                validate={function validate<T>(values: T) {
                    const MIN = 'should have minimun 3 characters'

                    const response = z
                        .object({
                            title: z.string().min(3, `Title ${MIN}`),
                            description: z
                                .string()
                                .min(3, `Description ${MIN}`),
                            dueDate: z.string(),
                        })
                        .safeParse(values)

                    return parseZodError(response)
                }}
                onSubmit={({ ...values }, helpers) => {
                    if (!currentUser) return
                    mutation.mutate({
                        values,
                        helpers,
                        currentUser: currentUser.uid,
                    })
                }}
            >
                {({ isSubmitting }) => {
                    return (
                        <Form noValidate>
                            <div className="flex flex-col gap-y-2 rounded-lg border border-gray-300 px-4 py-3 text-sm focus-within:border-gray-400">
                                <Field name="title">
                                    {({ meta, field }: FieldProps) => {
                                        const isInvalid =
                                            meta.touched && !!meta.error

                                        return (
                                            <input
                                                type="text"
                                                placeholder="Task name"
                                                className="outline-none"
                                                autoComplete="off"
                                                autoCapitalize="none"
                                                required
                                                disabled={isSubmitting}
                                                aria-invalid={isInvalid}
                                                aria-labelledby={
                                                    isInvalid
                                                        ? 'task-alert'
                                                        : undefined
                                                }
                                                {...field}
                                            />
                                        )
                                    }}
                                </Field>

                                <Field name="description">
                                    {({ meta, field }: FieldProps) => {
                                        const isInvalid =
                                            meta.touched && !!meta.error

                                        return (
                                            <input
                                                type="text"
                                                placeholder="Description"
                                                className="outline-none"
                                                autoComplete="off"
                                                autoCapitalize="none"
                                                required
                                                disabled={isSubmitting}
                                                aria-invalid={isInvalid}
                                                aria-labelledby={
                                                    isInvalid
                                                        ? 'task-alert'
                                                        : undefined
                                                }
                                                {...field}
                                            />
                                        )
                                    }}
                                </Field>

                                <section
                                    className="mt-2 flex gap-x-2"
                                    role="group"
                                    aria-label="Add Tags"
                                >
                                    <button
                                        type="button"
                                        className="inline-flex items-center gap-x-1 rounded border border-green-300 px-2.5 py-1.5 text-xs font-medium text-green-700 focus:outline-none focus-visible:ring-2 focus-visible:ring-green-300 focus-visible:ring-offset-2"
                                    >
                                        <MdCalendarToday aria-hidden />
                                        Today
                                    </button>

                                    <label className="relative inline-flex cursor-pointer items-center justify-center rounded-sm px-1.5 py-0.5 focus-within:ring-2 focus-within:ring-blue-300">
                                        <Field name="dueDate">
                                            {({
                                                field,
                                                form: { setFieldValue },
                                            }: FieldProps) => (
                                                <input
                                                    type={'date'}
                                                    className="focus:outline-none"
                                                    autoComplete="off"
                                                    autoCapitalize="none"
                                                    required
                                                    min={format(
                                                        new Date(),
                                                        'yyyy-MM-dd',
                                                    )}
                                                    {...field}
                                                />
                                            )}
                                        </Field>
                                    </label>

                                    <button
                                        type="button"
                                        className="inline-flex items-center gap-x-1 rounded border border-blue-300 px-2.5 py-1.5 text-xs font-medium text-blue-700 focus:outline-none focus-visible:ring-2 focus-visible:ring-blue-300 focus-visible:ring-offset-2"
                                    >
                                        <FaInbox
                                            aria-hidden
                                            className="text-base text-blue-500"
                                        />
                                        Inbox
                                    </button>
                                </section>
                            </div>

                            <div
                                role="alert"
                                className="text-xs capitalize leading-6 text-skin-main"
                                aria-live="polite"
                                id="task-alert"
                            >
                                {['title', 'description', 'dueDate'].map(
                                    (arg) => (
                                        <ErrorMessage name={arg} key={arg} />
                                    ),
                                )}
                            </div>

                            <div className="mt-2 flex justify-end gap-x-3 text-xsm">
                                <button
                                    type="button"
                                    className="rounded-sm bg-gray-300 px-3 py-1.5 font-medium text-gray-900 focus:outline-none focus-visible:ring-2 focus-visible:ring-gray-900 focus-visible:ring-offset-2"
                                    onClick={handleClose}
                                >
                                    Cancel
                                </button>

                                <button
                                    disabled={isSubmitting}
                                    type="submit"
                                    className="rounded-sm bg-skin-dark px-2 py-1.5 text-white focus:outline-none focus-visible:ring-2 focus-visible:ring-skin-main focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50"
                                >
                                    Update Task
                                </button>
                            </div>
                        </Form>
                    )
                }}
            </Formik>
        </>
    )
}

export default UpdateTask
