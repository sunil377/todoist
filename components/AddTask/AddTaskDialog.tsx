import { Dialog, Transition } from '@headlessui/react'
import { format } from 'date-fns'
import { ErrorMessage, Field, Form, Formik } from 'formik'
import { useKeypadListener } from 'hooks/useKeypadListener'
import { ITask } from 'index'
import { Fragment, useCallback, useMemo, useState } from 'react'
import { MdAdd } from 'react-icons/md'
import ToolTip from '../Tooltip'
import { useToCreateTask } from './hooks'
import ProjectPicker from './ProjectPicker'
import { validateTask } from './util'

function AddTaskDialog() {
    const [isOpen, setOpen] = useState(false)
    const handleCreateTask = useToCreateTask()

    const onkeydown = useCallback((e: KeyboardEvent) => {
        if (e.key === 'q') {
            e.preventDefault()
            setOpen((prev) => !prev)
        }
    }, [])

    useKeypadListener(onkeydown)
    const errorKeys: Array<keyof ITask> = useMemo(
        () => ['dueDate', 'project', 'description', 'title'],
        [],
    )

    return (
        <Fragment>
            <button
                className="tooltip relative rounded-full p-2 hover:bg-white/10 focus:outline-none focus-visible:bg-white/20"
                onClick={() => setOpen(true)}
            >
                <ToolTip className="top-full left-1/2 -translate-x-1/2 translate-y-2 after:-top-1 after:left-1/2 after:-translate-x-1/2">
                    Add Task <kbd>Q</kbd>
                </ToolTip>
                <MdAdd className="text-xl md:text-2xl" />
            </button>

            <Transition
                show={isOpen}
                enter="transition duration-100 ease-out"
                enterFrom="transform scale-95 opacity-0"
                enterTo="transform scale-100 opacity-100"
                leave="transition duration-75 ease-out"
                leaveFrom="transform scale-100 opacity-100"
                leaveTo="transform scale-95 opacity-0"
                as={Fragment}
            >
                <Dialog
                    onClose={setOpen}
                    className="fixed inset-0 z-50 flex items-center justify-center"
                >
                    <div
                        aria-hidden
                        className="fixed inset-0 z-10 bg-black/30"
                    />

                    <Dialog.Panel className="relative z-20 h-auto max-h-[calc(100vh-2*2.24rem)] w-full max-w-[min(700px,calc(100%-2rem))] overflow-auto rounded-md border border-gray-800 bg-gray-800 p-5 shadow-md shadow-gray-800 sm:p-10">
                        <Dialog.Title
                            as="h4"
                            className="mb-6 font-semibold tracking-wide text-white"
                        >
                            Add Task
                        </Dialog.Title>
                        <Formik
                            initialValues={{
                                title: '',
                                description: '',
                                dueDate: format(new Date(), 'yyyy-MM-dd'),
                                project: 'inbox',
                            }}
                            validate={validateTask}
                            onSubmit={(values, helpers) => {
                                handleCreateTask.mutate({
                                    values,
                                    helpers,
                                })
                            }}
                        >
                            {({
                                errors,
                                submitCount,
                                isSubmitting,
                                isValid,
                            }) => (
                                <Form noValidate className="space-y-3 text-sm">
                                    <div className="space-y-3">
                                        <Field
                                            type="text"
                                            name="title"
                                            autoComplete="off"
                                            autocurrect="off"
                                            className="w-full rounded border bg-white px-2 leading-8 focus:outline-none focus-visible:border-blue-300 focus-visible:ring focus-visible:ring-blue-300 focus-visible:ring-offset-2 focus-visible:ring-offset-gray-800"
                                            placeholder="Enter title..."
                                            aria-invalid={
                                                submitCount > 0 &&
                                                !!errors.title
                                            }
                                        />
                                        <Field
                                            type="text"
                                            name="description"
                                            autoComplete="off"
                                            autocurrect="off"
                                            className="w-full rounded border bg-white px-2 leading-8 focus:outline-none focus-visible:border-blue-300 focus-visible:ring focus-visible:ring-blue-300 focus-visible:ring-offset-2 focus-visible:ring-offset-gray-800"
                                            placeholder="Enter description..."
                                            aria-invalid={
                                                submitCount > 0 &&
                                                !!errors.description
                                            }
                                        />
                                    </div>
                                    <div className="flex gap-x-2 pt-4">
                                        <Field
                                            type="date"
                                            name="dueDate"
                                            autocurrect="off"
                                            autoComplete="off"
                                            autoCapitalize="none"
                                            required
                                            min={format(
                                                new Date(),
                                                'yyyy-MM-dd',
                                            )}
                                            className="rounded border bg-white px-2 leading-6 focus:outline-none focus-visible:border-blue-300 focus-visible:ring focus-visible:ring-blue-300 focus-visible:ring-offset-2 focus-visible:ring-offset-gray-800"
                                            aria-invalid={
                                                submitCount > 0 &&
                                                !!errors.dueDate
                                            }
                                        />
                                        <ProjectPicker isDialog />
                                    </div>
                                    <div className="flex justify-end gap-x-2 pt-2">
                                        <button
                                            type="button"
                                            className="rounded border border-gray-800 bg-transparent px-3 py-1.5 font-semibold text-white transition-colors hover:bg-white/10 focus:outline-none focus-visible:ring focus-visible:ring-blue-300 focus-visible:ring-offset-gray-800"
                                            onClick={() => setOpen(false)}
                                        >
                                            Cancel
                                        </button>
                                        <button
                                            type="submit"
                                            className="rounded bg-skin-main px-3 py-1.5 font-semibold text-white focus:outline-none focus-visible:ring focus-visible:ring-skin-main focus-visible:ring-offset-2 focus-visible:ring-offset-gray-800 disabled:pointer-events-none disabled:opacity-50"
                                            disabled={
                                                submitCount > 0 &&
                                                (!isValid || isSubmitting)
                                            }
                                        >
                                            Submit
                                        </button>
                                    </div>
                                    {submitCount > 0 &&
                                        errorKeys.map((er) => (
                                            <ErrorMessage
                                                name={er}
                                                key={er}
                                                component="span"
                                                className="block text-skin-main before:content-['*']"
                                            />
                                        ))}
                                </Form>
                            )}
                        </Formik>
                        <button
                            type="button"
                            className="absolute top-1.5 right-1.5 inline-flex h-8 w-8 items-center justify-center rounded-full bg-transparent text-xl leading-5 text-white transition-colors hover:bg-white/10"
                            onClick={() => setOpen(false)}
                        >
                            <span>&times;</span>
                        </button>
                    </Dialog.Panel>
                </Dialog>
            </Transition>
        </Fragment>
    )
}
export default AddTaskDialog
