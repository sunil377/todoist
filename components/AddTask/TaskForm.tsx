import { format, isValid } from 'date-fns'
import { ErrorMessage, Field, Form, FormikProps } from 'formik'
import { ITask } from 'index'
import ProjectPicker from './ProjectPicker'

function TaskForm({
    isSubmitting,
    submitCount,
    errors,
    close,
}: FormikProps<
    Pick<ITask, 'title' | 'description' | 'project'> & { dueDate: string }
> & {
    close: React.ReactNode
}) {
    return (
        <Form
            noValidate
            className="divide-y rounded-lg border border-gray-200 focus-within:border-gray-400"
        >
            <div>
                <div className="flex flex-col gap-y-2  px-4 py-3 text-sm">
                    <Field
                        name="title"
                        type="text"
                        placeholder="Enter title..."
                        className="outline-none"
                        autoComplete="off"
                        autoCapitalize="none"
                        required
                        autoFocus
                        disabled={isSubmitting}
                        aria-invalid={submitCount > 0 && !!errors.title}
                    />

                    <Field
                        name="description"
                        type="text"
                        placeholder="Enter description..."
                        className="outline-none"
                        autoComplete="off"
                        autoCapitalize="none"
                        required
                        autoFocus
                        disabled={isSubmitting}
                        aria-invalid={submitCount > 0 && !!errors.description}
                    />

                    <section
                        className="mt-2 flex gap-x-2 text-xsm"
                        role="group"
                        aria-label="Add Tags"
                    >
                        <Field
                            name="dueDate"
                            type="date"
                            className="cursor-pointer rounded-sm border border-green-300 px-2 leading-6 text-green-800 focus:outline-none focus-visible:border-green-500 focus-visible:ring-2"
                            autoComplete="off"
                            autoCapitalize="none"
                            required
                            min={format(new Date(), 'yyyy-MM-dd')}
                            aria-invalid={submitCount > 0 && !!errors.dueDate}
                        />

                        <ProjectPicker />
                    </section>
                </div>
            </div>

            <div className="p-2">
                <div
                    role="alert"
                    className="bg-black text-xs capitalize leading-6 text-skin-main"
                    aria-live="polite"
                    id="task-alert"
                >
                    {submitCount > 0 &&
                        ['title', 'description', 'dueDate', 'project'].map(
                            (arg) => <ErrorMessage name={arg} key={arg} />,
                        )}
                </div>

                <div className="flex justify-end gap-x-3 text-xsm">
                    {close}

                    <button
                        disabled={submitCount > 0 && (isSubmitting || !isValid)}
                        type="submit"
                        className="rounded border-4 border-transparent bg-skin-main px-2 py-1 text-white hover:bg-skin-dark focus:outline-none focus-visible:ring-2 focus-visible:ring-skin-main focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50"
                    >
                        Add task
                    </button>
                </div>
            </div>
        </Form>
    )
}
export default TaskForm
