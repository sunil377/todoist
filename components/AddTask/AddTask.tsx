import { Disclosure } from '@headlessui/react'
import { format } from 'date-fns'
import { Formik } from 'formik'
import { Fragment } from 'react'
import { MdOutlineAdd } from 'react-icons/md'
import { useToCreateTask } from './hooks'
import TaskForm from './TaskForm'
import { validateTask } from './util'

const INITIAL_VALUES = {
    title: '',
    description: '',
    dueDate: format(new Date(), 'yyyy-MM-dd'),
    project: 'inbox',
}

function RenderComponent({ close }: { close: React.ReactNode }) {
    const mutation = useToCreateTask()

    return (
        <>
            <Formik
                initialValues={INITIAL_VALUES}
                validate={validateTask}
                onSubmit={(values, helpers) => {
                    mutation.mutate({
                        values,
                        helpers,
                    })
                }}
            >
                {(formikprops) => {
                    return <TaskForm close={close} {...formikprops} />
                }}
            </Formik>
        </>
    )
}

function AddTask() {
    return (
        <Disclosure>
            {({ open }) => (
                <Fragment>
                    {!open && (
                        <Disclosure.Button className="group flex w-full items-center justify-start gap-x-2 rounded px-2 py-1 text-xsm focus:outline-none  focus-visible:ring-1 focus-visible:ring-indigo-300">
                            <span className="inline-flex items-center justify-center rounded-full text-skin-main transition-colors group-hover:bg-skin-main group-hover:text-white">
                                <MdOutlineAdd
                                    className="text-2xl"
                                    aria-hidden
                                />
                            </span>
                            <span className="font-normal text-gray-500 transition-colors group-hover:text-skin-dark">
                                Add Task
                            </span>
                        </Disclosure.Button>
                    )}

                    <Disclosure.Panel className="focus:outline-none">
                        {({ close }) => (
                            <RenderComponent
                                close={
                                    <button
                                        type="button"
                                        className="flex-shrink-0 rounded border-4 border-transparent px-2 py-1 font-medium text-gray-500 hover:text-gray-800"
                                        onClick={() => close()}
                                    >
                                        Cancel
                                    </button>
                                }
                            />
                        )}
                    </Disclosure.Panel>
                </Fragment>
            )}
        </Disclosure>
    )
}

export { AddTask as default, RenderComponent as AddTaskComponent }
