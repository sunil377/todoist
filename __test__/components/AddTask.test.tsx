import AddTask from '@/components/AddTask/AddTask'
import { render, screen } from '@testing-library/react'
import { QueryClient, QueryClientProvider } from 'react-query'

const client = new QueryClient()

const handleClose = jest.fn()

describe('AddTask.tsx', () => {
    test('render snapshot', () => {
        const { container } = render(
            <>
                <QueryClientProvider client={client}>
                    <AddTask onClose={handleClose} />
                </QueryClientProvider>
            </>,
        )

        expect(container).toMatchSnapshot()
    })

    test('textfield working fine', () => {
        render(
            <QueryClientProvider client={client}>
                <AddTask onClose={handleClose} />
            </QueryClientProvider>,
        )

        const taskInput = screen.getByPlaceholderText(/task/i)
        const descriptionInput = screen.getByPlaceholderText(/description/)
    })
})
