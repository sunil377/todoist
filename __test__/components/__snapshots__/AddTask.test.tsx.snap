// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTask.tsx render snapshot 1`] = `
<div>
  <h1>
    hello
  </h1>
</div>
`;
